### DISCUSSION TITLE
Sleigh Accumulator register usage

### DISCUSSION THREAD

**Discussion by lab313ru** [NationalSecurityAgency/ghidra#5289] [Q&A] [answered]
Hi,

I've created my own proc-module, and it includes hidden accumulator register usage. The problem is that the accumulator usage is just visible in a decompilation results:

![image](https://user-images.githubusercontent.com/7189309/236206195-3f5f0a66-8023-4c22-9757-a8ad70db559e.png)

Results in the 6502 module, for example don't contain acc's visible usage,

**Comment by lab313ru** [comment]
How can I hide such registers usage into a simple variables?

**Comment by astrelsky** [comment]
https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Processors/MIPS/data/languages/mips64.pspec#L20

I think this will do what you are looking for.

**Reply by lab313ru** [reply]
It didn't help.
```
<register name "acc" hidden="true"/>
```

**Reply by astrelsky** [reply]
> It didn't help.
> ```
> <register name "acc" hidden="true"/>
> ```
> 

There should be another setting in the pspec, something along the lines of registers appear directly in code. If I remember correctly there is a way to do what you're looking for I just can't seem to remember how.

**Reply by lab313ru** [reply]
> > It didn't help.
> > ```
> > <register name "acc" hidden="true"/>
> > ```
> 
> There should be another setting in the pspec, something along the lines of registers appear directly in code. If I remember correctly there is a way to do what you're looking for I just can't seem to remember how.

So, do you remember?)

**Reply by astrelsky** [reply]
> > > It didn't help.
> > > ```
> > > <register name "acc" hidden="true"/>
> > > ```
> > 
> > 
> > There should be another setting in the pspec, something along the lines of registers appear directly in code. If I remember correctly there is a way to do what you're looking for I just can't seem to remember how.
> 
> So, do you remember?)

I don't remember. It turns out I was thinking of addressesDoNotAppearDirectlyInCode.

**Reply by lab313ru** [reply]
> > > > It didn't help.
> > > > ```
> > > > <register name "acc" hidden="true"/>
> > > > ```
> > > 
> > > 
> > > There should be another setting in the pspec, something along the lines of registers appear directly in code. If I remember correctly there is a way to do what you're looking for I just can't seem to remember how.
> > 
> > 
> > So, do you remember?)
> 
> I don't remember. It turns out I was thinking of addressesDoNotAppearDirectlyInCode.

I believe that behavior happens after pseudo pcodes calling. Because the emulator don't know if an instruction changes acc or not.

**Reply by astrelsky** [reply]
My only remaining thought is that the decompiler normally wouldn't display a register like that unless it was declared as a global in the cspec or is not in the register address space. It's hard to tell what is really going on without looking at the implementation.

**Reply by lab313ru** [reply]
> My only remaining thought is that the decompiler normally wouldn't display a register like that unless it was declared as a global in the cspec or is not in the register address space. It's hard to tell what is really going on without looking at the implementation.

Declaration of the register is done like this:
```
define register offset=0x200 size=8 [acc sp pc ret];
```

And cspec only contains the following regarding registers (except calling conventions):
![image](https://github.com/NationalSecurityAgency/ghidra/assets/7189309/e932b15d-d74d-4cf5-989a-d3e67946bc1e)

**Reply by lab313ru** [reply]
> My only remaining thought is that the decompiler normally wouldn't display a register like that unless it was declared as a global in the cspec or is not in the register address space. It's hard to tell what is really going on without looking at the implementation.

And this is example of how such calls done:
```
:StaInArrayLiteral cReg,cReg2,[kIdx] is op=0x31; cReg; cReg2; kIdx {
    array_item_set(cReg, cReg2, acc);
}
```

**Reply by lab313ru** [reply]
> My only remaining thought is that the decompiler normally wouldn't display a register like that unless it was declared as a global in the cspec or is not in the register address space. It's hard to tell what is really going on without looking at the implementation.

I would like to send you what I have, but in some private way, if you allow.

**Reply by lab313ru** [reply, marked as answer]
Wow! I've removed register space from globals and it helped!

**Reply by lab313ru** [reply]
> My only remaining thought is that the decompiler normally wouldn't display a register like that unless it was declared as a global in the cspec or is not in the register address space. It's hard to tell what is really going on without looking at the implementation.

thanks a lot!

**Reply by astrelsky** [reply]
> Wow! I've removed register space from globals and it helped!

Yes I was just about to mention that. I'm not sure why it would even let you do that.
